fix: adds input restrictions for numeric fields

latin-panda opened the pull request:
Refactors string with appearance number question type

Closes #

### I have verified this PR works in these browsers (latest versions):

- [ ] Chrome
- [ ] Firefox
- [ ] Safari (macOS)
- [ ] Safari (iOS)
- [ ] Chrome for Android
- [ ] Not applicable

### What else has been done to verify that this works as intended?

### Why is this the best possible solution? Were any other approaches considered?

### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

### Do we need any specific form for testing your changes? If so, please attach one.

### What's changed

latin-panda commented on packages/web-forms/src/components/form-elements/input/InputNumeric.vue:
```diff
@@ -45,56 +52,29 @@ type NumericInputMode = 'decimal' | 'numeric';
 	inputMode = 'decimal';
 } else {
-	fractionalDigits = {};
+	fractionalDigits = { min: 0, max: 0 };
```
This resolves https://github.com/getodk/web-forms/issues/468

lognaturel commented:
It will be very nice to have this behavior aligned!

> Note that the user can still type the comma or dot, but once the field loses focus, the user moves to another question, or taps outside the field, these characters are removed. They are never passed to our components and are discarded by PrimeVue.

This one makes me uneasy. Is it possible to specify an input filter/restriction so that only digits can be typed at all? A user may not notice the change when leaving the field and going from "12.56" to "1256" is not great. I feel like these kinds of manipulations are acceptable when pasting because the user is looking at the field they're pasting into but it feels less ok on focus change.

latin-panda commented:
>  Is it possible to specify an input filter/restriction so that only digits can be typed at all?

It happens when double-tapping those keys. This is a related issue I haven't been able to fix: https://github.com/getodk/web-forms/issues/467

I want to try upgrading to PrimeVue first; otherwise, I'll attempt to work around it with keypress events.

lognaturel commented:
Sounds good, we can leave that issue open and dig deeper separately.

srujner commented:
@latin-panda

**Overall:** 

-Verified All Question Types, Thousands Separator and String Widgets form; 
- Tested if String type has a digit limitation; 
- Tested if Decimal type has a limit to 15 digits
- Tested if Integer type has a limit to 9 digits
- Tested if only numeric values can be entered and non-numeric characters are not accepted
- Tested very large numbers, negative numbers
- Tested with leading zeros to ensure that the field handle it properly (“0” is replaced by the next entered digit) 
- Issues:  https://github.com/getodk/web-forms/issues/263 https://github.com/getodk/web-forms/issues/415  https://github.com/getodk/web-forms/issues/416 https://github.com/getodk/web-forms/issues/468 
Are no longer reproducing
- I didn’t found any UI/UX issues on Iphone

**Iphone:**

- Looks like Holding slider button does not work (On web a can hold it and the number will increase faster, on mobile it is not supported) 
- Keyboard is displayed by default for number fields
- Keyboard on Iphone does not allow non-numeric inputs in
- Rotating the screen, blocking the phone does not introduced any issues


**Other:**

- In String Widget form thousand separator does not work for StringNumberWidget but I’m not sure if this is not a form issue
- On Iphone the slider button stays selected after increasing/decreasing number (it's hardly visible on screenshot) 
![IMG_8353](https://github.com/user-attachments/assets/3989dac3-4e28-4f71-ba53-72bfe2708b58)
- The screen "jumps" when the limit in the number field is reached and you click on the increase button
https://drive.google.com/file/d/1ExnGtEezIFZR96fDFF4GplnSFsuAHEo2/view?usp=sharing

latin-panda commented:
@srujner thanks for verifying and providing all those details, it's very useful! 

> In String Widget form thousand separator does not work for StringNumberWidget but I’m not sure if this is not a form issue

I didn't add support for that appearance, I'm working on that right now 🛠️

latin-panda commented:
@lognaturel @srujner I've added support for the "thousands-sep" in string type with "numbers" appearance.

I've updated the [test server](https://getodk.slack.com/archives/C051L6AKG4F/p1755104131688959) with the latest code.   

This PR is ready.

lognaturel reviewed (commented):
Wonderful to have all these addressed! Couple of tiny suggested changes inline.

Did y'all get a chance to try things in a locale where the thousands separator is a space, @srujner? French, for example. Edit: I see evidence of that in the screenshot at https://github.com/getodk/web-forms/pull/470#issuecomment-3188437886 so seems like yes

latin-panda commented:
> Did y'all get a chance to try things in a locale where the thousands separator is a space, @srujner?

Yes, here is a screenshot:
<img width="1119" height="1016" alt="Screenshot 2025-08-20 at 12 47 08 AM" src="https://github.com/user-attachments/assets/7357e905-7d7c-45c5-b27e-4970d9825d25" />

lognaturel commented:
Thanks!

Are the e2e tests currently running in CI? I'm not able to quickly confirm that they are.

latin-panda commented:
> Are the e2e tests currently running in CI? I'm not able to quickly confirm that they are.

Yes, and we added 10 new test cases 🤩, here for example:

<img width="1576" height="853" alt="Screenshot 2025-08-20 at 1 13 42 AM" src="https://github.com/user-attachments/assets/44a849c8-3bf3-45a5-84ae-5111acedad3c" />
